pallets/pips/src/lib.rs

```diff
@@ -315,15 +308,33 @@ pub struct SnapshotMetadata<T: Trait> {
 }
 
 /// A PIP in the snapshot's priority queue for consideration by the GC.
-#[derive(Encode, Decode, Clone, PartialEq, Eq)]
+#[derive(Encode, Decode, Copy, Clone, PartialEq, Eq)]
 #[cfg_attr(feature = "std", derive(Debug))]
-pub struct SnapshottedPip<T: Trait> {
+pub struct SnapshottedPip<Balance> {
     /// Identifies the PIP this refers to.
     pub id: PipId,
     /// Weight of the proposal in the snapshot's priority queue.
     /// Higher weights come before lower weights.
     /// The `bool` denotes the sign, where `true` siginfies a positive number.
-    pub weight: (bool, BalanceOf<T>),
+    pub weight: (bool, Balance),
+}
+
+/// Defines sorting order for PIP priority queues, with highest priority *last*.
+/// Having higher prio last allows efficient tail popping, so we have a LIFO structure.
+fn compare_spip<B: Ord + Copy>(l: &SnapshottedPip<B>, r: &SnapshottedPip<B>) -> Ordering {
+    let (l_dir, l_stake): (bool, B) = l.weight;
+    let (r_dir, r_stake): (bool, B) = r.weight;
+    l_dir
+        .cmp(&r_dir) // Negative has lower prio.
+        .then_with(|| match l_dir {
+            true => l_stake.cmp(&r_stake), // Higher stake, higher prio...
+            // Unless negative stake, in which case lower abs stake, higher prio.
+            false => r_stake.cmp(&l_stake),
+        })
+        // Lower id was made first, so assigned higher prio.
+        // This also gives us sorting stability through a total order.
+        // Moreover, as `queue` should be in by-id order originally.
+        .then(r.id.cmp(&l.id))
```

Comment:
Renaming `l` and `r` as something like `newer` and `older` would give a hint on the intended order of parameter application.

Reply:
I don't think `newer` and `older` are semantically right though, e.g., in the `on_runtime_upgrade` case where you want to recompute the entire live queue. And besides, parameter order should not matter as if you apply it in reverse, you'll get reverse results, so binary search and sorting should yield the same order.